core/src/main/scala/chisel3/internal/firrtl/IR.scala

```diff
@@ -326,10 +326,26 @@ private[chisel3] object ir {
     choices:    Seq[(String, BaseModule)])
       extends Definition
   case class DefObject(sourceInfo: SourceInfo, id: HasId, className: String) extends Definition
-  case class WhenBegin(sourceInfo: SourceInfo, pred: Arg) extends Command
-  case class WhenEnd(sourceInfo: SourceInfo, firrtlDepth: Int, hasAlt: Boolean = false) extends Command
-  case class AltBegin(sourceInfo: SourceInfo) extends Command
-  case class OtherwiseEnd(sourceInfo: SourceInfo, firrtlDepth: Int) extends Command
+
+  class When(val sourceInfo: SourceInfo, val pred: Arg) extends Command {
+    val ifRegion = new VectorBuilder[Command]
+    private var _elseRegion: Option[VectorBuilder[Command]] = None
+    def elseRegion: VectorBuilder[Command] = {
+      if (_elseRegion.isEmpty) {
+        _elseRegion = Some(new VectorBuilder[Command])
+      }
+      _elseRegion.get
+    }
+  }
+
+  object When {
+    def unapply(when: When): Option[(SourceInfo, Arg, Seq[Command], Seq[Command])] = {
+      Some(
+        (when.sourceInfo, when.pred, when.ifRegion.result(), when._elseRegion.map(_.result()).getOrElse(Seq.empty))
+      )
+    }
+  }
```

Comment:
```suggestion
    private var _elseRegion: VectorBuilder[Command] = null
    def elseRegion: VectorBuilder[Command] = {
      if (_elseRegion == null) {
        _elseRegion = new VectorBuilder[Command]
      }
      _elseRegion
    }

  object When {
    def unapply(when: When): Option[(SourceInfo, Arg, Seq[Command], Seq[Command])] = {
      Some(
        (when.sourceInfo, when.pred, when.ifRegion.result(), Option(when._elseRegion).map(_.result()).getOrElse(Seq.empty))
      )
    }
  }
```
`null` is awful in public APIs, but it saves 16-bytes per instance of `When` that has an has an `else`. Since it's only relevent in about 6 lines of code, it's worth it. Note that `Option(_)` boxing the `null` "reifies" it into the Option world safely, and because that's a function call, it can (and will in practice) be inlined and optimized away (unlike a field of type Option which cannot be optimized, or at least isn't by the Scala compiler today).

Reply:
Also consider `.map(_.result()).getOrElse(Seq.empty))` => `.fold(Seq.empty)(_.result())`. I prefer that but I respect others may not 🙂

Reply:
Good suggestion. Fixed in: https://github.com/chipsalliance/chisel/pull/4184/commits/9279b9ccef7ca754c91a7df19bbc6584940a28ab